raft/raft.go

```diff
@@ -381,21 +397,34 @@ func (r *raft) appendEntry(es ...pb.Entry) {
 // tickElection is run by followers and candidates after r.electionTimeout.
 func (r *raft) tickElection() {
 	if !r.promotable() {
-		r.elapsed = 0
+		r.electionElapsed = 0
 		return
 	}
-	r.elapsed++
+	r.electionElapsed++
 	if r.isElectionTimeout() {
-		r.elapsed = 0
+		r.electionElapsed = 0
 		r.Step(pb.Message{From: r.id, Type: pb.MsgHup})
 	}
 }
 
 // tickHeartbeat is run by leaders to send a MsgBeat after r.heartbeatTimeout.
 func (r *raft) tickHeartbeat() {
-	r.elapsed++
-	if r.elapsed >= r.heartbeatTimeout {
-		r.elapsed = 0
+	r.heartbeatElapsed++
+	r.electionElapsed++
+
+	if r.electionElapsed >= r.electionTimeout {
+		r.electionElapsed = 0
+		if r.checkQuorum {
```

Comment:
If the `checkQuorum` var is true, we will do the checking by sending a msg to state machine or we will not do the checking at all.

The input of the statemachine MUST be a message.